pulsar-client-cpp/lib/ProducerImpl.cc

```diff
@@ -726,10 +716,10 @@ Future<Result, ProducerImplBaseWeakPtr> ProducerImpl::getProducerCreatedFuture()
 uint64_t ProducerImpl::getProducerId() const { return producerId_; }
 
 void ProducerImpl::handleSendTimeout(const boost::system::error_code& err) {
-    Lock lock(mutex_);
     if (state_ != Pending && state_ != Ready) {
```

Comment:
```suggestion
    const auto state = state_;
    if (state != Pending && state != Ready) {
```

Not sure if the compiler will optimize it (read the same value for the two `state_ != xxx` comparisons). But it's better to get the current state only once explicitly.